### PR TITLE
feature:add workspace, message and conversation schema

### PR DESCRIPTION
**Pull request by avinashankur**
## Description
PR primarily focused on renaming "channel" to "workspace" across the project and adding new models.

### Renaming and Refactoring
- Renamed **Channel** to **Workspace** across the codebase.
- Updated the **Member Controller** to handle changes due to the renaming.
- Adjusted the **Member Model** to reflect the transition from "channel" to "workspace."

### New Models and Controllers
- Added **Channel Model** and its corresponding controller.
- Introduced the following models:
  - **Conversation Model**
  - **Message Model**
  - **Workspace Model**

### User Enhancements
- Added **Avatar** support in the **User Model**.

### Routing
- Added **Workspace Routes** to enable workspace-related functionalities.

### CODE REVIEW

File: controllers/auth.controller.js [hunk 1]
```diff
@@ -33,9 +34,21 @@ const registerUser = asyncHandler(async (req, res, next) => {
       .status(409)
       .json({ message: "User with email or username already exists" });
   }
+
```
Comment (HarshSharma0801) [inline controllers/auth.controller.js:37]: why this is required ? in this function ?

File: models/conversation.model.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        ref: "Message",
+      },
+    ],
+    memberId: [
+      {
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "User",
+      },
+    ],
+    channelId: [
```
Comment (HarshSharma0801) [inline models/conversation.model.js:22]: why this is an array , each channel will contain a single conversation

File: models/conversation.model.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      enum: ["DM", "Channel"],
+      required: true,
+    },
+    messages: [
+      {
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Message",
+      },
+    ],
+    memberId: [
```
Comment (HarshSharma0801) [inline models/conversation.model.js:16]: here why ref is from User when Member modal is present , take a look at design which i sent

File: models/message.model.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+const messageSchema = new mongoose.Schema(
+  {
+    content: {
+      type: String,
+    },
+    attachment: {
+      type: String,
+    },
+    memberId: [
```
Comment (HarshSharma0801) [inline models/message.model.js:11]: again why ref is User , and why its an array a message will be sent by a single member so its unique

File: models/message.model.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import mongoose from "mongoose";
+
+const messageSchema = new mongoose.Schema(
+  {
+    content: {
+      type: String,
+    },
+    attachment: {
```
Comment (HarshSharma0801) [inline models/message.model.js:8]: attachments will be array as multiple files can be there

**Review by HarshSharma0801 — CHANGES_REQUESTED** [review]

File: controllers/auth.controller.js (hunk 1 above)
Comment (AbhinavG786) [reply controllers/auth.controller.js:37]: Sir, existed user isme check nahi karenge kya agar naya user register karna chah raha toh ki woh email already present hai database mein?

File: controllers/channel.controller.js [hunk 6]
```diff
@@ -1,35 +1,27 @@
...
-            headers: {
-                Authorization: `Client-ID ${process.env.UNSPLASH_ACCESS_KEY}`,
-            },
-            params: {
-                count: 1,
-            },
-        });        
-        const randomImage = response.url;
+        const { workspaceId } = req.params
+        const { name, userId } = req.body
```
Comment (HarshSharma0801) [inline controllers/channel.controller.js:13]: make checks if these ids and name is not present

File: controllers/channel.controller.js [hunk 7]
```diff
@@ -78,17 +70,14 @@ const getChannelUsingId = async (req, res, next) => {
 const updateChannel = async (req, res, next) => {
     try {
         const { id } = req.params;
-        const { name, description } = req.body;
+        const { name } = req.body;
```
Comment (HarshSharma0801) [inline controllers/channel.controller.js:73]: again checks

File: controllers/member.controller.js [hunk 8]
```diff
@@ -1,86 +1,95 @@
...
         logger.error(error.message);
         return res.status(500).json({ message: "Error adding member", error });
     }
 };
 
-//To find and return the complete channel and user object about all the channels a user is part of.
-const getChannelMembers = async (req, res, next) => {
+//To find and return the complete workspace and user object about all the workspaces a user is part of.
+const getWorkspaceMembers = async (req, res, next) => {
     const { userId } = req.params;
```
Comment (HarshSharma0801) [inline controllers/member.controller.js:37]: same as above

File: controllers/member.controller.js [hunk 9]
```diff
@@ -1,86 +1,95 @@
...
+            return res.status(404).json({ message: 'This user is not part of any Workspace' });
         }
         const response = await Promise.all(members.map(async (member) => {
-            const channel = await Channel.findById(member.channelId).populate('owner');
+            const workspace = await Workspace.findById(member.workspaceId).populate('owner');
             const user = await User.findById(member.userId);
             return {
-                channel: channel,
+                workspace: workspace,
                 user: user,
```
Comment (HarshSharma0801) [inline controllers/member.controller.js:49]: why sending user is required here , just asking

File: controllers/workspace.controller.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
+import { Workspace } from "../models/workspace.model.js"
```
[the remark this answers is not in the document]
Comment (HarshSharma0801) [reply controllers/workspace.controller.js:1]: maintain spacings line by line also

File: controllers/channel.controller.js [hunk 11]
```diff
@@ -1,35 +1,27 @@
...
         const newChannel = new Channel({
             name,
             members: [],
-            description,
-            avatar: randomImage,
+            workspaceId: workspaceId,
             owner: userId
         })
         const saveChannel = await newChannel.save()
         const OwnerMember = new Members({
```
Comment (HarshSharma0801) [inline controllers/channel.controller.js:21]: you can add spaces here where there is a different db operation

File: controllers/member.controller.js (hunk 9 above)
Comment (AbhinavG786) [reply controllers/member.controller.js:49]: Sir ,actually this is the member api we have made changes in after renaming channel to workspace.You had asked to send the channel and user object in the member get apis(in google doc) so earlier I had send the user object here to tell what other workspaces the user is part of.

**Pull request merged**